tools/setup-envtest/versions/types.go

```diff
@@ -0,0 +1,229 @@
+
+package versions
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+)
+
+// Platform contains OS & architecture information
+// Either may be '*' to indicate a wildcard match.
+type Platform struct {
+	OS   string
+	Arch string
+}
+
+// Matches indicates if this platform matches the other platform,
+// potentially with wildcard values.
+func (p Platform) Matches(other Platform) bool {
+	return (p.OS == other.OS || p.OS == "*" || other.OS == "*") &&
+		(p.Arch == other.Arch || p.Arch == "*" || other.Arch == "*")
+}
+
+// IsWildcard checks if either OS or Arch are set to wildcard values.
+func (p Platform) IsWildcard() bool {
+	return p.OS == "*" || p.Arch == "*"
+}
+func (p Platform) String() string {
+	return fmt.Sprintf("%s/%s", p.OS, p.Arch)
+}
+
+// PlatformItem represents a platform with corresponding
+// known metadata for its download.
+type PlatformItem struct {
+	Platform
+	MD5 string
+}
+
+const (
+	// AnyPoint matches any point version.
+	AnyPoint PointVersion = -1
+	// LatestPoint matches any point version, but indicates we
+	// want the latest available.
+	LatestPoint PointVersion = -2
+)
+
+// LatestVersion requests the most recent available version.
+var LatestVersion = Version{
+	Major: LatestPoint,
+	Minor: LatestPoint,
+	Patch: LatestPoint,
+}
+
+// PointVersion represents a major, minor, or patch version,
+// or potentially a wildcard.
+type PointVersion int
+
+// IsWildcard indicates if this will match any version.
+func (p PointVersion) IsWildcard() bool {
+	return p == AnyPoint || p == LatestPoint
+}
+
+// Matches checks if two point versions are compatible.
+// Two point versions are compatible if they are
+// a) both concrete
+// b) one is a wildcard, and the other is not
+//
+// Comparing two wildcards will panic.
+func (p PointVersion) Matches(other PointVersion) bool {
+	switch {
+	case !p.IsWildcard() && !other.IsWildcard():
+		return p == other
+	case p.IsWildcard() && other.IsWildcard():
+		panic("tried to compare two wildcard point versions")
+	default: // a wildcard matches with any concrete version
+		return true
+	}
+}
+func (p PointVersion) String() string {
+	switch p {
+	case AnyPoint:
+		return "*"
+	case LatestPoint:
+		return "+"
+	default:
+		return strconv.Itoa(int(p))
+	}
+}
+
+// PointVersionFromValidString extracts a point version
+// from the corresponding string representation, which may
+// be a number >= 0, "+" (LatestPoint), or "*" (AnyPoint).
+//
+// Anything else will cause a panic (use this on strings
+// extracted from regexes).
+func PointVersionFromValidString(str string) PointVersion {
+	switch str {
+	case "+":
+		return LatestPoint
+	case "*":
+		return AnyPoint
+	default:
+		ver, err := strconv.Atoi(str)
+		if err != nil {
+			panic(err)
+		}
+		return PointVersion(ver)
+	}
+}
+
+// Version is an X.Y.Z version.
+type Version struct {
+	Major, Minor, Patch PointVersion
+}
+
+// IsWildcard checks if any part of this version is a wildcard.
+func (v Version) IsWildcard() bool {
```

Comment:
I wanted a quick way to distinguish "get me the latest" from "get me anything from disk if I have it, otherwise download", but you may be right that it's easier.  I'll play around a bit and see if I can find something that works nicely

Reply:
also, FWIW: all semver selectors are bespoke as far as I can tell -- there doesn't appear to be a spec for the selectors, just the version itself.